### PR TITLE
Explicitly transfer target height of an Enrollment

### LINKED ISSUE #2838: First enrollment of a node can associate a PreImage with wrong height
`checkAndEnroll` creates a new enrollment and sends it to the network, since there is no previous enrollment nodes accept it regardless of it's commitment.

If externalization of the enrollment gets delayed for even a single block, the commitment in the enrollment is associated with the wrong height, causing any PreImage revealed after that to be deemed invalid.

### PR DESCRIPTION
**Pull request by omerfirmak**
Previously, target height was assumed to be the next height. After
next block is externalized, enrollment was removed from the pool, even
if not externalized.

If an Enrollment gets gossiped after externalization of it's target
height that the originating node intended, it could be externalized
at a later height. Which caused the commitment in the Enrollment to be
associated with a wrong height and made impossible to reveal any new
PreImages.

By sending target height explicitly along the enrollment, we guarantee
that a delay in externalization of Enrollments don't doom the Validator.

Fixes #2838

### CODE REVIEW

**Review by hewison-chris — APPROVED** [review]
This works but would it not be better to just add this height to the `Enrollment` struct?

**Review by Geod24 — CHANGES_REQUESTED** [review]
This doesn't really fix the issue: If a byzantine validator changes the value and gossip it faster than us, we could end up in the same situation (and get slashed).
This kind of issues is why we originally had enrollment cycles. Adding the height to the Enrollment was something we tried very hard to avoid in the past.

**Comment by omerfirmak** [comment]
> : If a byzantine validator changes the value and gossip it faster than us, we could end up in the same situation (and get slashed)

Yep, thanks for pointing that out.

> Adding the height to the Enrollment was something we tried very hard to avoid in the past.

Let's just sign the intended height as well but not add it to the struct, nodes will try to validate Enrollment with the assumption that it is intended for the next height.

So `enroll_sig = hashMulti(height, enrollment).sign();`

**Comment by Geod24** [comment]
That's slick, I like it a lot.

File: source/agora/api/FullNode.d [hunk 1]
```diff
@@ -322,7 +322,7 @@ public interface API
 
     ***************************************************************************/
 
-    public void postEnrollment (in Enrollment enroll);
+    public void postEnrollment (in Enrollment enroll, in Height avail_height);
```
Comment (linked0) [inline source/agora/api/FullNode.d:332]: Needs documentation for the `avail_height`.

File: source/agora/node/FullNode.d [hunk 2]
```diff
@@ -1048,10 +1048,14 @@ public class FullNode : API
...
 
     /// POST /enrollment
-    public override void postEnrollment (in Enrollment enroll) @safe
+    public override void postEnrollment (in Enrollment enroll, in Height avail_height) @safe
     {
         this.recordReq("postEnrollment");
 
+        // Ignore enrollments targeting a previous Height
+        if (avail_height <= this.ledger.getBlockHeight())
+            return;
```
Comment (linked0) [inline source/agora/node/FullNode.d:1067]: It would be better to add a log for the ignoring.

**Comment by omerfirmak** [comment]
> So enroll_sig = hashMulti(height, enrollment).sign();

Added a second commit implementing that, we may squash it before merging

**Review by linked0 — APPROVED** [review]
Great but the second commit could be split into two commits.

**Comment by omerfirmak** [comment]
> Great but the second commit could be split into two commits.

I thought of it as pretty atomic, how should I do that?

File: source/agora/consensus/EnrollmentManager.d [hunk 3]
```diff
@@ -819,48 +818,42 @@ unittest
...
-    // Adding enrollments in reverse order of utxo order to show it still works
-    const EnrollAt10 = Height(10);
-    ordered_enrollments.sort!("a.utxo_key > b.utxo_key");
-    foreach (idx, ordered_enroll; ordered_enrollments)
-        assert(man.addEnrollment(ordered_enroll, pairs[idx].address, EnrollAt10,
-            &utxo_set.peekUTXO, getPenaltyDeposit));
-    enrolls = man.getEnrollments(EnrollAt10, &utxo_set.peekUTXO, getPenaltyDeposit);
-    assert(enrolls.length == 3);
-    assert(enrolls.isStrictlyMonotonic!("a.utxo_key < b.utxo_key"));
-
```
Comment (linked0) [inline source/agora/consensus/EnrollmentManager.d:832]: I thought that this is unrelated to this commit.

File: source/agora/consensus/EnrollmentManager.d [hunk 4]
```diff
@@ -898,41 +891,37 @@ unittest
 
     Height height = Height(2);
 
-    Enrollment[] enrollments;
-    foreach (idx, kp; pairs[0 .. 3])
-    {
-        enrollments ~= EnrollmentManager.makeEnrollment(
-            utxo_hashes[idx], kp, height, params.ValidatorCycle);
-    }
-
```
Comment (linked0) [inline source/agora/consensus/EnrollmentManager.d:907]: Ditto.

**Review by linked0 — COMMENTED** [review]
I spotted the unrelated modification to the second commit. But it could be thought of as atomic based on the point of view. Good job, go ahead as it is.

File: source/agora/consensus/EnrollmentManager.d (hunk 4 above)
Comment (omerfirmak) [reply source/agora/consensus/EnrollmentManager.d:907]: Those are both required to make the tests pass with this change.

**Comment by omerfirmak** [comment]
Added a simple test as well

**Review by Geod24 — APPROVED** [review]

**Comment by Geod24** [comment]
Can you rebase ?

**Comment by omerfirmak** [comment]
Rebased

**Pull request merged**